### PR TITLE
[Torchax][Refactor] Consolidate weight loading logics

### PR DESCRIPTION
**Pull request by kyuyeunk**
# Description

No change in functionality. Pure refactoring changes. Main motivation is to remove duplicated logics in different quantization backend & consolidate them into a single layer specific file / function.

- move/rename `fused_moe.py` 
  - rename it into `fused_moe_gmm.py` to distinguish it from all-kernel based approach.
  - move from `tpu_inference/layers/vllm/fused_moe.py` to `tpu_inference/layers/common/fused_moe_gmm.py` since it's pure functional logic that doesn't have any dependencies with pytorch
- create folder `tpu_inference/layers/vllm/process_weights`
  - Move all layer specific weight processing logics into this folder
    - create & move all fused moe related weight processing logic into `tpu_inference/layers/vllm/process_weights/fused_moe_weights.py`
    - create & moved all linear related weight processing logic into `tpu_inference/layers/vllm/process_weights/linear_weights.py`
    - delete duplicated weight loading logics in each quantization backends (like `awq.py`, `mxfp4.py`) and use the consolidated weight processing logic in the newly created files
  - moved all weight sharding related logics that is not covered by above two files into `tpu_inference/layers/vllm/process_weights/cleanup_sharding.py`
- create interface for different fused_moe bakcends 
  - create `tpu_inference/layers/vllm/fused_moe.py` (different from now renamed `fused_moe_gmm.py`) 
  - create enum `FusedMoEBackend` that allows specifying which backend the code is using
  - create forward function that invokes different backends (gmm based vs. all kernel based) depending on the argument
  - remove duplicated backend routing logic in each quantization backends and use the consolidated forward function instead
- create `tpu_inference/layers/common/utils.py`
  - move logics in `tpu_inference/layers/vllm/linear_common.py` that is shared by both linear and fused moe layers into the util file.
- config file changes
  - rename `tpu_inference/layers/vllm/quantization/common.py` into `tpu_inference/layers/vllm/quantization/configs.py` for better clarity
  - rename `JaxCommonLinearConfig` into `VllmQuantLinearConfig`
  - rename `JaxCommonConfig` into `VllmQuantConfig`

# Tests

No functionality change. PR is good if all unit test passes.

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/layers/vllm/quantization/unquantized.py [hunk 1]
```diff
@@ -94,35 +90,50 @@ def get_quant_method(self, layer: torch.nn.Module,
...
-                self.jax_config.n_shards,
-                self.jax_config.fuse_matmuls,
-            )
+            bias = t2j(layer.bias, use_dlpack=False)
             delattr(layer, "bias")
-            layer.bias = bias
+        else:
+            bias = None
+
+        # TODO: Use jit to speedup weight loading.
```
Comment (jrplatin) [inline tpu_inference/layers/vllm/quantization/unquantized.py]: To be addressed here?

File: tpu_inference/layers/vllm/quantization/mxfp4.py [hunk 2]
```diff
@@ -140,271 +145,74 @@ def process_weights_after_loading(self, layer: torch.nn.Module):
...
         w2_bias = t2j(layer.w2_bias, use_dlpack=False)
 
-        # Wrap functions in jit to speedup requantization.
         @jax.jit
         def wrapper(w13_weight, w13_weight_scale, w13_bias, w2_weight,
                     w2_weight_scale, w2_bias):
             # Dequantize fp4 weights into fp32.
             w13_weight = dequantize_tensor_from_mxfp4_packed(
-                w13_weight, w13_weight_scale, 2)
+                w13_weight, w13_weight_scale)
```
Comment (jrplatin) [inline tpu_inference/layers/vllm/quantization/mxfp4.py]: Just want to confirm -- the `dequantize_tensor_from_mxfp4_packed` function signature changed here right, hence why you removed the `2`?

File: tpu_inference/layers/vllm/quantization/unquantized.py [hunk 3]
```diff
@@ -94,35 +90,50 @@ def get_quant_method(self, layer: torch.nn.Module,
...
+                None,
+                None,
+                bias,
+                fused=self.linear_config.fuse_matmuls,
+                output_sizes=self.linear_config.output_sizes,
+                reorder_size=self.linear_config.n_shards,
+            )
+
+        weights = wrapper(weight, bias)
+        weight, _, _, bias = torch_view(
```
Comment (jrplatin) [inline tpu_inference/layers/vllm/quantization/unquantized.py]: Just curious -- why do we need to call `t2j` on `weights` above and then also call this?

File: tpu_inference/layers/vllm/quantization/unquantized.py [hunk 4]
```diff
@@ -182,247 +194,89 @@ def _apply_split(self,
...
+        mesh: Mesh,
+        ep_axis_name: str = "model",
+    ):
+        super().__init__(moe)
+        self.mesh = mesh
+        self.moe_backend = select_moe_backend(self.moe)
+
+        self.extra_backend_kwargs = {}
+        if self.moe_backend == FusedMoEBackend.FUSED_MOE:
+            self.extra_backend_kwargs = dict(
```
Comment (jrplatin) [inline tpu_inference/layers/vllm/quantization/unquantized.py:217]: Might be nice to add a comment explaining why this is needed since it's not immediately obvious to me

File: tpu_inference/layers/vllm/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_int8.py [hunk 5]
```diff
@@ -38,55 +43,65 @@
...
-            )
+            bias = t2j(layer.bias, use_dlpack=False)
             delattr(layer, "bias")
-            layer.bias = bias
+        else:
+            bias = None
+
+        # TODO: Use jit to speedup weight loading.
+        # @jax.jit
+        def wrapper(weight, weight_scale, bias):
```
Comment (jrplatin) [inline tpu_inference/layers/vllm/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_int8.py]: I think I saw a similar helper method elsewhere in the PR -- maybe consolidate to a utils file?

Comment (jrplatin) [reply tpu_inference/layers/vllm/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_int8.py]: Also, maybe update the name the be more descriptive

File: tpu_inference/layers/vllm/process_weights/fused_moe_weights.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+from tpu_inference.layers.common.utils import \
+    reorder_concatenated_tensor_for_sharding
+from tpu_inference.layers.vllm.fused_moe import FusedMoEBackend
+from tpu_inference.utils import align_to
+
+P = PartitionSpec
+N = NamedSharding
+
+
+def quantize_moe_weights(w13_weight: jax.Array, w13_bias: jax.Array | None,
```
Comment (jrplatin) [inline tpu_inference/layers/vllm/process_weights/fused_moe_weights.py]: Can you add docstrings / return hints? Thanks! Same with the rest of the PR/file

File: tpu_inference/layers/vllm/process_weights/fused_moe_weights.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+    return w13_weight, w13_weight_scale, w13_bias, w2_weight, w2_weight_scale, w2_bias
+
+
+def shard_moe_weights(weights: list[jax.Array | None],
+                      *,
+                      moe_backend: FusedMoEBackend,
+                      mesh: Mesh,
+                      per_channel=False):
+    # TODO: Use proper class structure.
```
Comment (jrplatin) [inline tpu_inference/layers/vllm/process_weights/fused_moe_weights.py]: To be adressed here?

File: tpu_inference/layers/vllm/process_weights/fused_moe_weights.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+    match moe_backend:
+        case FusedMoEBackend.FUSED_MOE | FusedMoEBackend.GMM_EP:
+            ep_sharding = N(mesh, P(ShardingAxisName.EXPERT))
+            weight_shardings = [ep_sharding] * 6
+        case FusedMoEBackend.GMM_TP:
+            w2_weight_scale_p_spec = P() if per_channel else P(
+                None,
+                ShardingAxisName.MLP_TENSOR,
+            )
+            weight_shardings = [
```
Comment (jrplatin) [inline tpu_inference/layers/vllm/process_weights/fused_moe_weights.py]: Can you add what each of the dims here is correspond to? It's hard to catch from first glance, thanks!

File: tpu_inference/layers/vllm/process_weights/linear_weights.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+    return weight, weight_scale, zero_point, bias
+
+
+def shard_linear_weights(weights: list[jax.Array | None],
+                         *,
+                         mesh: Mesh,
+                         weight_p_spec: PartitionSpec,
+                         bias_p_spec: PartitionSpec,
+                         transposed: bool = True):
+    # TODO: Use proper data structure.
```
Comment (jrplatin) [inline tpu_inference/layers/vllm/process_weights/linear_weights.py]: To be addressed here?

File: tpu_inference/layers/vllm/quantization/awq.py [hunk 10]
```diff
@@ -74,80 +78,79 @@ def get_quant_method(
...
-            )
+            bias = t2j(layer.bias, use_dlpack=False)
             delattr(layer, "bias")
-            layer.bias = bias
+        else:
+            bias = None
+
+        # TODO: Use jit to speedup weight loading.
+        # @jax.jit
+        def wrapper(weight, weight_scale, zero_point, bias):
```
Comment (jrplatin) [inline tpu_inference/layers/vllm/quantization/awq.py]: Same comment as elsewhere -- does it make sense to move this to a common utils file / consolidate this with similar functions elsewhere?

File: tpu_inference/layers/vllm/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py [hunk 11]
```diff
@@ -71,15 +79,74 @@ def __init__(
...
-        jax_config: JaxCommonLinearConfig,
+        linear_config: VllmQuantLinearConfig,
     ):
         super().__init__(weight_quant, is_static_input_scheme)
 
-        self.jax_config = jax_config
+        self.linear_config = linear_config
 
     def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
+        # TODO: Move scale processing logic into a consolidated file.
```
Comment (jrplatin) [inline tpu_inference/layers/vllm/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py]: To be addressed here?

File: tpu_inference/layers/vllm/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py [hunk 12]
```diff
@@ -71,15 +79,74 @@ def __init__(
...
+        if layer.bias is not None and not layer.skip_bias_add:
+            if layer.return_bias:
+                logger.warning_once("Bias might return incorrect value.")
+            bias = t2j(layer.bias, use_dlpack=False)
+            delattr(layer, "bias")
+        else:
+            bias = None
+
+        # TODO: Use jit to speedup weight loading.
+        # @jax.jit
```
Comment (jrplatin) [inline tpu_inference/layers/vllm/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py]: Same as above

File: tpu_inference/layers/vllm/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py [hunk 13]
```diff
@@ -71,15 +79,74 @@ def __init__(
...
+            )
+
+        shard_weights = functools.partial(
+            shard_linear_weights,
+            mesh=self.linear_config.mesh,
+            weight_p_spec=self.linear_config.weight_sharding,
+            bias_p_spec=self.linear_config.bias_sharding,
+        )
+
+        # If per-tensor quantization is used, skip processing for scale.
```
Comment (jrplatin) [inline tpu_inference/layers/vllm/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py]: Why?

File: tpu_inference/layers/vllm/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 14]
```diff
@@ -83,17 +84,26 @@ def get_moe_method(
...
-                                            JaxCommonConfig):
+                                            VllmQuantConfig):
 
     def __init__(
         self,
         weight_quant: QuantizationArgs,
         input_quant: QuantizationArgs,
         moe: FusedMoEConfig,
         mesh: Mesh,
+        sharding_axis: str = "model",
```
Comment (jrplatin) [inline tpu_inference/layers/vllm/quantization/compressed_tensors/compressed_tensors_moe.py]: Is it safe to hardcode this? I doubt it will ever really change on the JAX side, but just came to mind

File: tpu_inference/layers/vllm/process_weights/fused_moe_weights.py (hunk 8 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/process_weights/fused_moe_weights.py]: done.

File: tpu_inference/layers/vllm/quantization/mxfp4.py (hunk 2 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/quantization/mxfp4.py]: The default behavior is to dequantize along last dim - which is the 2nd dim. so strictly speaking, there isn't a need to pass which dimension we should dequantize over. but I've added it back to make it more explicit.

File: tpu_inference/layers/vllm/process_weights/fused_moe_weights.py (hunk 7 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/process_weights/fused_moe_weights.py]: done.

File: tpu_inference/layers/vllm/process_weights/fused_moe_weights.py (hunk 6 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/process_weights/fused_moe_weights.py]: done.

File: tpu_inference/layers/vllm/process_weights/linear_weights.py (hunk 9 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/process_weights/linear_weights.py]: done.

File: tpu_inference/layers/vllm/quantization/unquantized.py (hunk 1 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/quantization/unquantized.py]: in an actual workload where we process multiple weights with same shapes, i found jax.jit is faster since compile time overhead is negliglble since it's an one time cost. but in unit test where we run the layer through multiple different shapes, i found that it significantly slows donw the ci.

to fix that, i need to refactor unit test as well. I was hoping to make this pr (relatively) small, but let me fix the unit test as well & address the comment now.

File: tpu_inference/layers/vllm/quantization/unquantized.py (hunk 3 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/quantization/unquantized.py]: jax.jit can only accept jax array (or torchax array), but weights are initially a pytorch array. so we need to do a conversion from pytorch array to jax array (or torchax) before we can pass it to jax.jit funciton.

File: tpu_inference/layers/vllm/quantization/awq.py (hunk 10 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/quantization/awq.py]: No, what i've observed was that each quantization methods may require special processing logic. For instance, mxfp4 requires requantizing into fp4 block weight and awq requires some special unpacking logic.

so my idea is to let `process_lienar_weights` and `process_moe_weights` to be a common shared function. and before calling that function, each quantization methods can perform some logic that is specific to their checkpoint.

alternative approach is to move these specific logic into `process_linear_weights` as well use some branching logic to handle them, but that would defeat the single-responsibility principle

File: tpu_inference/layers/vllm/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py (hunk 11 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py]: done.

File: tpu_inference/layers/vllm/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py (hunk 13 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py]: modified the code so this isn't true anymore.

File: tpu_inference/layers/vllm/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py (hunk 12 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py]: done.

File: tpu_inference/layers/vllm/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 14 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/quantization/compressed_tensors/compressed_tensors_moe.py]: We need to make this configurable, but since this PR is pure refactoring without any additiona of features (like making some argument configurable), I'll leave it as-is.

File: tpu_inference/layers/vllm/quantization/unquantized.py (hunk 4 above)
Comment (kyuyeunk) [reply tpu_inference/layers/vllm/quantization/unquantized.py:217]: done.
